src/index.js

```diff
@@ -10,113 +10,207 @@ const RETRIABLE_ERRORS = [
   // ignore server sent html error pages
   // or truncated json responses
   'SyntaxError',
-]
+];
 
-module.exports = createInfuraMiddleware
-module.exports.fetchConfigFromReq = fetchConfigFromReq
+module.exports = createInfuraMiddleware;
+module.exports.fetchConfigFromReq = fetchConfigFromReq;
 
-function createInfuraMiddleware (opts = {}) {
-  const network = opts.network || 'mainnet'
-  const maxAttempts = opts.maxAttempts || 5
-  const { source, projectId, headers = {} } = opts
+/**
+ * Arguments to {@link fetch}.
+ *
+ * @typedef {object} FetchConfig
+ * @property {string} fetchUrl - The URL to fetch.
+ * @property {object} fetchParams - The options object to pass to {@link fetch}.
+ * @property {string} fetchParams.method - The request method.
+ * @property {Record<string, string>} fetchParams.headers - The request headers.
+ * @property {string} fetchParams.body - The request body.
+ */
+
+/**
+ * Builds {@link json-rpc-engine}-compatible middleware designed for interfacing
+ * with Infura's JSON-RPC endpoints.
+ *
+ * @param {object} [opts] - The options (default: {}).
+ * @param {string} [opts.network] - A network that Infura supports; plugs into
+ * `https://${network}.infura.io` (default: 'mainnet').
+ * @param {number} [opts.maxAttempts] - The number of times a request to Infura
+ * should be retried in the case of failure (default: 5).
+ * @param {string} [opts.source] - A descriptor for the entity making the
+ * request; tracked by Infura for analytics purposes.
+ * @param {string} [opts.projectId] - The Infura project id.
```

Comment:
In practice this is not optional. An error is thrown if it is missing.

This makes the default value for `opts` pointless, as it isn't valid. So `opts` may as well be documented as required as well.

Reply:
Thanks! Fixed in cfaa406add3465d78c85749108c8a8d798b3abb4.